compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseLogicalObjectFifoIntoWorkgroup.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-amdaie-fuse-logical-objectfifo-into-workgroup"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// This pass fuses `LogicalObjectFifoFromMemref` operations into the
+/// workgroups where they are used. This is a utility which can be relied upon
+/// by other passes to perform custom transformations with logical objectfifos
+/// within a workgroup. This pass handles multiple workgroups and uses outside
+/// workgroups by cloning `LogicalObjectFifoFromMemref` into every workgroup
+/// where it's used.
+class AMDAIEFuseLogicalObjectFifoIntoWorkgroupPass
+    : public impl::AMDAIEFuseLogicalObjectFifoIntoWorkgroupBase<
+          AMDAIEFuseLogicalObjectFifoIntoWorkgroupPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+
+  AMDAIEFuseLogicalObjectFifoIntoWorkgroupPass() = default;
+  AMDAIEFuseLogicalObjectFifoIntoWorkgroupPass(
+      const AMDAIEFuseLogicalObjectFifoIntoWorkgroupPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIEFuseLogicalObjectFifoIntoWorkgroupPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  ModuleOp moduleOp = getOperation();
+  IRRewriter rewriter(context);
+
+  SmallVector<AMDAIE::LogicalObjectFifoFromMemrefOp> logicalObjectFifos;
+  moduleOp->walk([&](AMDAIE::LogicalObjectFifoFromMemrefOp op) {
+    if (isa<AMDAIE::WorkgroupOp>(op->getParentOp())) {
+      return WalkResult::advance();
+    }
+    logicalObjectFifos.push_back(op);
+    return WalkResult::advance();
```

Comment:
I dont think this advance is needed.

Reply:
It's needed because of the `if` block before. Otherwise, it will error on 

> non-void lambda does not return a value in all control paths [-Werror,-Wreturn-type]